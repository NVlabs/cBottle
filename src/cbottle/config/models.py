# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import dataclasses
import json


@dataclasses.dataclass
class ModelConfigV1:
    architecture: str = "unet_hpx64"
    model_channels: int = 128
    label_dim: int = 0
    out_channels: int = 1
    condition_channels: int = 0
    time_length: int = 1
    label_dropout: float = 0.0
    level: int = 6

    # arguments for SongUnetHPX1024
    position_embed_channels: int = 20
    img_resolution: int = 128

    def dumps(self):
        return json.dumps(dataclasses.asdict(self))

    @classmethod
    def loads(cls, s):
        return cls(**json.loads(s))
