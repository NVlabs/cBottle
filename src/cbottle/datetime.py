# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import datetime
import cftime
import numpy as np
import pandas as pd


def as_pydatetime(time) -> datetime.datetime:
    if isinstance(time, cftime.datetime):
        # very important to set the timezone to UTC for example when using timestamps
        return datetime.datetime(*cftime.to_tuple(time), tzinfo=datetime.timezone.utc)
    elif isinstance(time, datetime.datetime):
        return time
    else:
        raise NotImplementedError(type(time))


def as_numpy(time) -> np.ndarray:
    # Standardize time to np.ndarray of np.datetime64
    if hasattr(time, "values"):  # Handle pandas Index
        time = time.values
    elif isinstance(time, (pd.Timestamp, datetime.datetime)):
        time = np.array([np.datetime64(time)])
    elif isinstance(time, np.datetime64):
        time = np.array([time])
    else:
        time = np.array([np.datetime64(t) for t in time])
    return time


def second_of_day(time):
    begin_of_day = time.replace(hour=0, second=0, minute=0)
    return (time - begin_of_day).total_seconds()


def as_cftime(timestamp):
    return cftime.DatetimeGregorian(
        timestamp.year,
        timestamp.month,
        timestamp.day,
        timestamp.hour,
        timestamp.minute,
        timestamp.second,
    )
