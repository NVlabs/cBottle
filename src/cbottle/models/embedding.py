# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import torch
import math


class PositionalEmbedding(torch.nn.Module):
    """Timestep embedding used in the DDPM++ and ADM architectures.


    f = (1/M)^(i / N)
    [cos(f_i x), sin(f_i x)] for i =0,...,N - 1

    sup wavelength = sup  2 pi / f = 2 pi 1 / inf f = 2 pi / (1 / M) = 2 pi M

    """

    def __init__(self, num_channels, max_positions=10000, endpoint=False):
        super().__init__()
        self.num_channels = num_channels
        self.max_positions = max_positions
        self.endpoint = endpoint

    def forward(self, x):
        freqs = torch.arange(
            start=0, end=self.num_channels // 2, dtype=torch.float32, device=x.device
        )
        freqs = freqs / (self.num_channels // 2 - (1 if self.endpoint else 0))
        freqs = (1 / self.max_positions) ** freqs
        x = x.ger(freqs.to(x.dtype))
        x = torch.cat([x.cos(), x.sin()], dim=1)
        return x


class FourierEmbedding(torch.nn.Module):
    """Timestep embedding used in the NCSN++ architecture."""

    def __init__(self, num_channels, scale=16):
        super().__init__()
        self.register_buffer("freqs", torch.randn(num_channels // 2) * scale)

    def forward(self, x):
        x = x.ger((2 * math.pi * self.freqs).to(x.dtype))
        x = torch.cat([x.cos(), x.sin()], dim=1)
        return x


class FrequencyEmbedding(torch.nn.Module):
    """Periodic Embedding.

    Useful for inputs defined on the circle [0, 2pi)
    """

    def __init__(self, num_channels):
        super().__init__()
        self.register_buffer(
            "freqs", torch.arange(1, num_channels + 1), persistent=False
        )

    def forward(self, x):
        freqs = self.freqs[None, :, None, None]
        x = x[:, None, :, :]
        x = x * (2 * math.pi * freqs).to(x.dtype)
        x = torch.cat([x.cos(), x.sin()], dim=1)
        return x


class CalendarEmbedding(torch.nn.Module):
    """Time embedding assuming 365.25 day years

    Args:
        day_of_year: (n, t)
        second_of_day: (n, t)
    Returns:
        (n, embed_channels * 4, t, x)

    """

    def __init__(self, lon, embed_channels: int, include_legacy_bug: bool = False):
        """
        Args:
            include_legacy_bug: Provided for backwards compatibility
                with existing checkpoints. If True, use the incorrect formula
                for local_time (hour - lon) instead of the correct formula (hour + lon)
        """
        super().__init__()
        self.register_buffer("lon", lon, persistent=False)
        self.embed_channels = embed_channels
        self.embed_second = FrequencyEmbedding(embed_channels)
        self.embed_day = FrequencyEmbedding(embed_channels)
        self.out_channels = embed_channels * 4
        self.include_legacy_bug = include_legacy_bug

    def forward(self, day_of_year, second_of_day):
        if second_of_day.shape != day_of_year.shape:
            raise ValueError()

        if self.include_legacy_bug:
            local_time = (second_of_day.unsqueeze(2) - self.lon * 86400 // 360) % 86400
        else:
            local_time = (second_of_day.unsqueeze(2) + self.lon * 86400 // 360) % 86400

        a = self.embed_second(local_time / 86400)
        doy = day_of_year.unsqueeze(2)
        b = self.embed_day((doy / 365.25) % 1)
        a, b = torch.broadcast_tensors(a, b)
        return torch.concat([a, b], dim=1)  # (n c x)
