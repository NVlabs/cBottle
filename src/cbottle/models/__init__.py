# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import torch

from cbottle.models import networks
from cbottle.config.models import ModelConfigV1


def get_model(
    config: ModelConfigV1,
    *,
    use_apex_groupnorm: bool | None = None,
    padding_backend=None,
    in_place_operations: bool = True,
) -> torch.nn.Module:
    if config.architecture == "unet_hpx64":
        precond_cls = networks.EDMPrecond
        if config.time_length > 1:
            architecture = networks.SongUNetHPX64Video(
                config.out_channels + config.condition_channels,
                config.out_channels,
                label_dim=config.label_dim,
                calendar_embed_channels=8,
                model_channels=config.model_channels,
                time_length=config.time_length,
                label_dropout=config.label_dropout,
                calendar_include_legacy_bug=config.calendar_include_legacy_bug,
                use_apex_groupnorm=use_apex_groupnorm,
                padding_backend=padding_backend,
                in_place_operations=in_place_operations,
            )
        else:
            architecture = networks.SongUNetHPX64(
                config.out_channels + config.condition_channels,
                config.out_channels,
                label_dim=config.label_dim,
                calendar_embed_channels=8,
                model_channels=config.model_channels,
                calendar_include_legacy_bug=config.calendar_include_legacy_bug,
                enable_classifier=config.enable_classifier,
                use_apex_groupnorm=use_apex_groupnorm,
                padding_backend=padding_backend,
                in_place_operations=in_place_operations,
            )

    elif config.architecture == "unet_hpx1024_patch":
        precond_cls = networks.EDMPrecondLegacy
        architecture = networks.SongUnetHPXPatch(
            in_channels=config.condition_channels
            + config.out_channels
            + config.position_embed_channels,
            out_channels=config.out_channels,
            img_resolution=config.img_resolution,
            model_channels=config.model_channels,
            pos_embed_channels=config.position_embed_channels,
            label_dim=config.label_dim,
            level=config.level,
            use_apex_groupnorm=use_apex_groupnorm,
            padding_backend=padding_backend,
            in_place_operations=in_place_operations,
        )
    else:
        raise NotImplementedError(config.architecture)

    return precond_cls(
        model=architecture,
        domain=architecture.domain,
        img_channels=config.out_channels,
        time_length=config.time_length,
        label_dim=config.label_dim,
    )
