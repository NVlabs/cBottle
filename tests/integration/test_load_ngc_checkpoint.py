# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from datetime import datetime
from earth2studio.data import CBottle3D
from earth2studio.models.dx import CBottleSR


def test_e2s_cbottle_coarse(regtest):
    package = CBottle3D.load_default_package()
    ds = CBottle3D.load_model(package).to("cuda")
    cbottle_da = ds([datetime(2022, 9, 5)], ["msl", "tcwv"])
    cbottle_da.to_dataset(name="sample").info(regtest)


def test_e2s_cbottle_super_load():
    package = CBottleSR.load_default_package()
    super_resolution_window = (
        0,
        -120,
        50,
        -40,
    )  # (lat south, lon west, lat north, lon east)
    CBottleSR.load_model(
        package,
        output_resolution=(1024, 1024),
        super_resolution_window=super_resolution_window,
    )
