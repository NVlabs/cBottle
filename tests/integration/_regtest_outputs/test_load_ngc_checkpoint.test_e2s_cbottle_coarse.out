xarray.Dataset {
dimensions:
	time = 1 ;
	variable = 2 ;
	lat = 721 ;
	lon = 1440 ;

variables:
	datetime64[ns] time(time) ;
	<U4 variable(variable) ;
	float64 lat(lat) ;
	float64 lon(lon) ;
	float64 sample(time, variable, lat, lon) ;

// global attributes:
}